### DISCUSSION TITLE
去水印后仍然有轻微痕迹（注意了水印背景与原图尺寸保持一致）

### DISCUSSION THREAD

**Discussion by Water1573** [lanyeeee/bilibili-manga-watermark-remover#47] [Q&A] [answered]
### 在提问之前

- [x] 我已经查看了[常见问题](https://github.com/lanyeeee/bilibili-manga-watermark-remover?tab=readme-ov-file#%E5%B8%B8%E8%A7%81%E9%97%AE%E9%A2%98)，我确定这个问题不在其中

### 问题描述

去水印后仍然有轻微痕迹（注意了水印背景与原图尺寸保持一致）
因为漫画部分章节无法自动生成背景水印，所以就拿了同本漫画不同章节之前生成了同尺寸的背景水印图直接放入背景水印文件夹中
发现能自动生成水印背景的章节能正常去水印（无残留），但无法自动生成背景的（或无法同时生成和白两张背景的）就会有残留（不想“常见问题”中b漫添加了多重水印的情况，残留颜色太淡了）
<img width="1123" alt="image" src="https://github.com/user-attachments/assets/52cf53f4-8d8d-46a0-bdf5-7a5c1d019cae">
<img width="914" alt="image" src="https://github.com/user-attachments/assets/2e8335bf-c8a5-47db-9353-48183942a2b1">

### 预期行为

正常去除水印，且无残留

### 实际行为

无法自动生成水印背景的图即使导入相同尺寸的背景图，也会有水印去除不干净的问题（具体流程有点多，我在评论里详细说）

### 截图或录屏

<img width="914" alt="image" src="https://github.com/user-attachments/assets/0dad8e35-89a6-434c-b813-7772315d56c6">

### 工具版本号

v0.6

### 其他

本人之前尝试过将多个章节图片放在一个文件夹中处理，但奇怪的是仍然是能自动生成完整背景图的章节能够正常去除水印，反之其他章节一样会有残留（颜色都很淡），由于无法导出运行日志，所以就没贴了T-T

**Comment by Water1573** [comment]
由于图片文件和背景文件有点大，我想用百度网盘的方式贴出来，望佬能谅解T-T
链接：https://pan.baidu.com/s/1SVqkS6k_zkAdX3fP2hj_qg?pwd=21rX 
提取码：21rX

**Comment by lanyeeee** [comment, marked as answer]
先说结论，这是正常现象，无论是001话还是210话，包括你贴出来的图都已经成功去水印了

下面是001话图11去` 水印后` 和` 对去水印后的图片调色` 的效果
|去水印|调色|
|-|-|
|![image](https://github.com/user-attachments/assets/2eef4702-98b3-4fd8-b00a-cbe9107a8b0a)|![image](https://github.com/user-attachments/assets/6ccb4ebd-8a2f-4320-90a5-506df485cc12)| 

不难发现，即使成功去水印了也是有轻微残留的，只是肉眼看不出来

原因比较复杂，先说结论，是JPEG压缩算法导致的
为了保证大家都能看懂，我把图像处理的过程类比为解方程

1. 原图: 100
2. 对原图加水印: 100 * x = 500
3. 带水印的原图: 500
4. 去水印: 500 -> 100

不难看出，即使知道加水印算法是`原图 * x`，我们要完成 `500 -> 100`，还需要知道 `x`的值
现在知道为什么要`黑/白背景水印图`了吗，就是用来算`x`的，每种水印，每部漫画，每个尺寸的`x`都可能不同

当 `加水印算法` 和 `x` 的值都集齐后，理论上就能完成 `500 -> 100` 了
然而因为JPEG压缩算法的缘故，它会轻微改变`带水印的原图(500)`的值，所以实际的过程其实是下面这样

1. 原图: 100
2. 对原图加水印: 100 * x = 500
3. 带水印的原图: 500
4. 对带水印的原图用JPEG压缩: 500 -> 495
5. 带水印的被JPEG压缩的图片: 495
6. 去水印: 495 -> 99

这就是 `即使成功去水印了也是有轻微残留` 的原因


2025年2月6日追加
刚才用v0.8.0(去水印算法优化后的版本)试了一下，即使调色也看不到残留了
|v0.8.0去水印|v0.8.0调色|
|-|-|
|![image](https://github.com/user-attachments/assets/1cb7deca-23c3-44f2-b436-01bef577559c)|![image](https://github.com/user-attachments/assets/a9eec4bf-e5b0-487a-94ff-a10fd0bab7c9)|
